### PR TITLE
Enable various sampling techniques

### PR DESCRIPTION
**Pull request by bocchris-aws**
Enabling multiple sampling techniques by wrapping the models forward function. Before each forward pass we pad the input and move respective tensors to the XLA device. Directly after the forward, we "unpad" the logits such that any subsequent processing will work. This way we can call the forward pass on the XLA device without recompilations. 
Tested on T5-small and BART-base for constrained beam search.

### CODE REVIEW

**Comment by bocchris-aws** [comment]
@michaelbenayoun Would you mind taking a look?

**Comment by michaelbenayoun** [comment]
Yes, but will wait for the release to merge

**Comment by chrisby** [comment]
@michaelbenayoun Happy to incorporate any feedback when you find the time.

**Comment by chrisby** [comment]
@michaelbenayoun Ping

**Comment by michaelbenayoun** [comment]
Is it ready to be merged?

**Comment by bocchris-aws** [comment]
> Is it ready to be merged?

Now it is.

File: optimum/neuron/generation/utils.py [hunk 1]
```diff
@@ -1080,7 +1080,9 @@ def generate(
             )
 
         else:
-            # 10.a go into not optimized generation modes
+            # 10.a go into not optimized generation modes. We can't handle KV cache in this case.
+            assert model_kwargs['use_cache'] == False, "`use_cache` is currently not supported for this sampling method."
```
Comment (michaelbenayoun) [inline optimum/neuron/generation/utils.py]: Can we reword that as an exception please?
We do not do asserts in the main code in `optimum-neuron`.

Comment (bocchris-aws) [reply optimum/neuron/generation/utils.py]: Yes, changed it to raising ValueError intead.

**Comment by michaelbenayoun** [comment]
Can you re-trigger the CI please, a lot was failing.
